src/main/java/org/folio/rest/impl/EholdingsProxyTypesImpl.java

```diff
@@ -1,97 +1,47 @@
-
-  private final Logger logger = LoggerFactory.getLogger(EholdingsProxyTypesImpl.class);
-
-  @Autowired
-  private RMAPIConfigurationService configurationService;
   @Autowired
-  private ProxyConverter converter;
-  @Autowired
-  private HeaderValidator headerValidator;
+  private RMAPITemplateFactory rmapiTemplateFactory;
 
   public EholdingsProxyTypesImpl() {
     SpringContextUtil.autowireDependencies(this, Vertx.currentContext());
   }
 
-  public EholdingsProxyTypesImpl(RMAPIConfigurationService configurationService, ProxyConverter converter,
-    HeaderValidator headerValidator) {
-    this.configurationService = configurationService;
-    this.converter = converter;
-    this.headerValidator = headerValidator;
-  }
-
   @Override
   @HandleValidationErrors
   public void getEholdingsProxyTypes(Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
-    headerValidator.validate(okapiHeaders);
-
-    CompletableFuture.completedFuture(null)
-      .thenCompose(o -> configurationService.retrieveConfiguration(new OkapiData(okapiHeaders)))
-      .thenCompose(rmapiConfiguration -> {
-        RMAPIService rmapiService = new RMAPIService(rmapiConfiguration.getCustomerId(), rmapiConfiguration.getAPIKey(),
-          rmapiConfiguration.getUrl(), vertxContext.owner());
-
-        return rmapiService.retrieveProxies();
-      })
-      .thenAccept(proxies -> {
-        ProxyTypes proxyTypes = converter.convert(proxies);
-
-        asyncResultHandler.handle(succeededFuture(respond200WithApplicationVndApiJson(proxyTypes)));
-      })
-      .exceptionally(e -> {
-        logger.error("Failed to get proxy types", e);
-
-        if(e.getCause() instanceof RMAPIUnAuthorizedException){
-          RMAPIUnAuthorizedException rmApiException = (RMAPIUnAuthorizedException)e.getCause();
-
-          asyncResultHandler.handle(Future.succeededFuture(GetEholdingsProxyTypesResponse
-            .status(HttpStatus.SC_FORBIDDEN)
-            .header(CONTENT_TYPE_HEADER, JSON_API_TYPE)
-            .entity(createError(rmApiException.getMessage()))
-            .build()));
-        }
-        else {
-          asyncResultHandler.handle(Future.succeededFuture(GetEholdingsProxyTypesResponse
-            .status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
-            .header(CONTENT_TYPE_HEADER, JSON_API_TYPE)
-            .entity(createError(e.getCause().getMessage()))
-            .build()));
-        }
-        return null;
-      });
+    rmapiTemplateFactory.createTemplate(okapiHeaders, asyncResultHandler)
+      .requestAction((rmapiService, okapiData) ->
+        rmapiService.retrieveProxies()
+      )
+      .addErrorMapper(RMAPIUnAuthorizedException.class, rmApiException -> GetEholdingsProxyTypesResponse
+        .status(HttpStatus.SC_FORBIDDEN)
+        .header(CONTENT_TYPE_HEADER, JSON_API_TYPE)
+        .entity(createError(rmApiException.getMessage()))
+        .build())
```

Comment:
I think in general we could get an RMAPIUnAuthorizedException when calling any of the rm api related endpoints. It was explicitly added for this ProxyTypes endpoint but I'm wondering if the handling of this could be more generalized in the RMAPITemplate class?

Reply:
This is probably more responsibility of `ErrorHandler`. It already contains some default logic:
![image](https://user-images.githubusercontent.com/42244720/50837441-4f52e200-1364-11e9-80d4-08ae429d47f0.png)

If RMAPIUnAuthorizedException is common to all methods and should be handled in the same way, default mapper seems like a right place to do it

Reply:
thanks @carolegodfrey 
I'll generalize the mapping

Reply:
Thanks @dmtkachenko

Reply:
refactored